language-plutus-core/src/Language/PlutusCore/DeBruijn.hs

```diff
@@ -0,0 +1,266 @@
+{-# LANGUAGE DerivingStrategies    #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+-- | Support for using de Bruijn indices for term and type names.
+module Language.PlutusCore.DeBruijn (
+                                Ix (..)
+                                , DeBruijn (..)
+                                , TyDeBruijn (..)
+                                , FreeVariableError (..)
+                                , deBruijnTy
+                                , deBruijnTerm
+                                , deBruijnProgram
+                                , unDeBruijnTy
+                                , unDeBruijnTerm
+                                , unDeBruijnProgram
+                                ) where
+
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Pretty
+import           Language.PlutusCore.Quote
+import           Language.PlutusCore.Type
+
+import           Control.Exception
+import           Control.Lens               hiding (Level, index, ix)
+import           Control.Monad.Except
+import           Control.Monad.Reader
+
+import qualified Data.Bimap                 as BM
+import qualified Data.Text                  as T
+import           Data.Typeable
+
+import           Numeric.Natural
+
+import           GHC.Generics
+
+-- | A relative index used for de Bruijn identifiers.
+newtype Ix = Ix Natural
```

Comment:
Not wild about stealing names from `base`.

Reply:
Huh, I didn't know there was an `Ix` in base. I'll use `Index` which doesn't seem to be claimed by anything important.